src/System.Net.Security/tests/FunctionalTests/NegotiateStreamInvalidOperationTest.cs

```diff
@@ -0,0 +1,274 @@
+            using (var serverStream = new VirtualNetworkStream(network, isServer: true))
+            using (var client = new NegotiateStream(clientStream))
+            using (var server = new NegotiateStream(serverStream))
+            {
+                Task[] auth = new Task[2];
+                auth[0] = client.AuthenticateAsClientAsync();
+                auth[1] = server.AuthenticateAsServerAsync();
+                await TestConfiguration.WhenAllOrAnyFailedWithTimeout(auth);
+
+                server.Dispose();
+                Assert.Throws<IOException>(() => client.Write(_sampleMsg, 0, _sampleMsg.Length));
+                Assert.Throws<IOException>(() => client.Read(recvBuf, 0, _sampleMsg.Length));
+            }
+        }
+
+        [Fact]
+        public async Task NegotiateStream_InvalidParametersForReadWrite_Throws()
+        {
+            VirtualNetwork network = new VirtualNetwork();
+            byte[] buffer = _sampleMsg;
+            int offset = 0;
+            int count = _sampleMsg.Length;
+
+            using (var clientStream = new VirtualNetworkStream(network, isServer: false))
+            using (var serverStream = new VirtualNetworkStream(network, isServer: true))
+            using (var client = new NegotiateStream(clientStream))
+            using (var server = new NegotiateStream(serverStream))
+            {
+                // Need to do authentication first, because Read/Write operation
+                // is only allowed using a successfully authenticated context.
+                Task[] auth = new Task[2];
+                auth[0] = client.AuthenticateAsClientAsync(CredentialCache.DefaultNetworkCredentials, string.Empty);
+                auth[1] = server.AuthenticateAsServerAsync();
+                await TestConfiguration.WhenAllOrAnyFailedWithTimeout(auth);
+
+                // Null buffer.
+                Assert.Throws<ArgumentNullException>(() => client.Write(null, offset, count));
+
+                // Negative offset.
+                Assert.Throws<ArgumentOutOfRangeException>(() => client.Write(buffer, -1, count));
+
+                // Negative count.
+                Assert.Throws<ArgumentOutOfRangeException>(() => client.Write(buffer, offset, -1));
+
+                // Invalid offset and count combination.
+                Assert.Throws<ArgumentOutOfRangeException>(() => client.Write(buffer, offset, count + count));
+
+                // Null buffer.
+                Assert.Throws<ArgumentNullException>(() => server.Read(null, offset, count));
+
+                // Negative offset.
+                Assert.Throws<ArgumentOutOfRangeException>(() => server.Read(buffer, -1, count));
+
+                // Negative count.
+                Assert.Throws<ArgumentOutOfRangeException>(() => server.Read(buffer, offset, -1));
+
+                // Invalid offset and count combination.
+                Assert.Throws<ArgumentOutOfRangeException>(() => server.Read(buffer, offset, count + count));
```

Comment:
Do any of these argument exceptions have argument names that we can validate in these asserts?